### DISCUSSION TITLE
c3 and object oriented.

### DISCUSSION THREAD

**Discussion by saimmm** [c3lang/c3c#1608] [General]
Is there any suggestion or planification about c3 language to be object oriented  language?

**Comment by lerno** [comment]
No there isn't.

**Comment by funny-falcon** [comment]
Given there are interfaces and dynamic methods, I'd call it "object-oriented", since polyphormism and incapsulation are provided with this mechanics.

**Reply by lerno** [reply]
You're of course permitted to call it that, but it lacks static methods and variables, which breaks normal OO patterns, and the stdlib explicitly opts out of classic `object.do_stuff()` code.

**Reply by funny-falcon** [reply]
I don't get how static methods and variables bound to OO patterns.

yes, stdlib is not OO :-) but could be.

**Reply by goyalyashpal** [reply]
> interfaces and dynamic methods, I'd call it "object-oriented", since polyphormism and incapsulation
> \- _@funny-falcon at https://github.com/c3lang/c3c/discussions/1608#discussioncomment-12006567_

> [c3] lacks static methods and variables, which breaks normal OO patterns
> \- _@lerno at https://github.com/c3lang/c3c/discussions/1608#discussioncomment-12228853_

[anyinterfaces]: https://c3-lang.org/generic-programming/anyinterfaces/ "Interfaces and Any Type | C3"

There's a distinction:
* Object Based (OB): structures, functions, modules
* Object Oriented (OO): OB with inheritance, & specialization

As for polymorphism, if you think about:
- it has no relation to thing being object or not.  
- type-based polymorhpism like [anyinterfaces] would not be "Object Oriented" either.

Examples:
- The julia language with its multiple dispatch based on parameter type and still not being OO is good example of that.  
- The C-styled C++ with templates & name-mangling based type polymorphism is another example.

^

- I learnt this term Object Based in a course, and it felt so great to know about the name for majority of chasm existing between strictly "Procedural" vs "Object Oriented".
- This may not make sense now, but it does when you look at the evolution of the capabilities of early prog languages and how their design was strictly constrained by capabilities of computers at that time.

It feels like a controversy that this name and this concept was obliterated from the general discourse.  As if some people would loose unique selling point of their language if the counterpart to which they are trying to prove superior to, is similar sounding paradigm as well.

**Reply by goyalyashpal** [reply]
[SO]: https://stackoverflow.com/q/15430004/a#31601869 "java - Core difference between object oriented and object based language"
[wiki]: https://en.wikipedia.org/wiki/Object-based_language "Object-based language"

Found some resources on OBP vs OOP:
* Answer on StackOverflow: [15430004/a#31601869][so]
* Article on wikipedia: [object-based language][wiki]
